test/services/purchase_stock_test.rb

```diff
@@ -0,0 +1,48 @@
+require "test_helper"
+
+# Q: Can transaction cash balance go negative?
+
+class PurchaseStockTest < ActiveSupport::TestCase
+  setup do
+    @student = users(:one)
+  end
+
+  test "it creates a withdrawl transaction in portfolio_transactions" do
+    order = Order.create(stock: Stock.first, shares: 5, status: :pending, user: @student)
+    assert_difference("PortfolioTransaction.count") do
+      PurchaseStock.execute(order)
+    end
+
+    portfolio_transaction = PortfolioTransaction.last
+    assert_equal "withdrawal", portfolio_transaction.transaction_type
+    assert_equal portfolio_transaction, order.portfolio_transaction
+    assert_operator portfolio_transaction.amount, :<, 0
+  end
+
+  test "it creates an linked entry in portfolio_stocks" do
+    order = Order.create(stock: Stock.first, shares: 5, status: :pending, user: @student)
+    assert_difference("PortfolioStock.count") do
+      PurchaseStock.execute(order)
+    end
+    portfolio_stock = PortfolioStock.last
+    assert_equal portfolio_stock, order.portfolio_stock
+  end
+
+  test "it updates the order status to completed" do
```

Comment:
Do we have time this weekend to test any failure states?

Reply:
Added a failure state at the bottom, required `mocha` gem for stubbing (seemed like a popular way in minitest? I tend to use rspec)

Reply:
I wonder if there's a way we can do this without the stub you've written. I don't want to hold up the good work in this PR, though. If you're up for taking a go at rewriting it, I'm happy to pair with you to go over my concerns. If you're not up for that, then I think we should merge this PR in and we can discuss putting follow-up work in a separate issue

Reply:
I won't have availability for a bit so feel free to do what seems best. The mocking + failure test are contained in the last commit so you could remove that before merging if you don't want mocha added

Reply:
I've been noodling on this. For the kind of fail cases we want to test, are you talking about something like "what if there is not enough money in the account" or something like that @h-m-m ?

Reply:
Yes! I think that's the main situation.

We will run into a problem at some point where the pending orders plus the new order could add up to more than what's in the account. Should that also go in this PR, or should we break that out into a follow-up PR?

Reply:
I'm glad you asked this because it made me realize some of this logic should maybe live in a more traditional controller request. Can't we process the transactions themselves based on the price the user sees when they make the request. I would think the only part we need to put in a chron job is updating the price of the stock, either at the beginning or end of the trade day. I'm thinking the purchases themselves should just use that number so the user can see a more immediate error message if they don't have enough in their account. Am I making sense? Is there another reason to put the transactions themselves in a job?